app/scripts/migrations/077-supplements/077-supplement-for-088.ts

```diff
@@ -0,0 +1,161 @@
+import { hasProperty, Hex, isObject, isStrictHexString } from '@metamask/utils';
+import { BN } from 'ethereumjs-util';
+
+/**
+ * Deletes properties of `NftController.allNftContracts`, `NftController.allNfts`,
+ * `TokenListController.tokensChainsCache`, `TokensController.allTokens`,
+ * `TokensController.allIgnoredTokens` and `TokensController.allDetectedTokens` if
+ * their keyed by decimal number chainId and another hexadecimal chainId property
+ * exists within the same object.
+ * Further explanation in ./077-supplements.md
+ *
+ * @param state - The persisted MetaMask state, keyed by controller.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export default function transformState077For086(
+  state: Record<string, unknown>,
+): Record<string, unknown> {
+  if (hasProperty(state, 'NftController') && isObject(state.NftController)) {
+    const nftControllerState = state.NftController;
+
+    // Migrate NftController.allNftContracts
+    if (
+      hasProperty(nftControllerState, 'allNftContracts') &&
+      isObject(nftControllerState.allNftContracts)
+    ) {
+      const { allNftContracts } = nftControllerState;
+
+      if (
+        Object.keys(allNftContracts).every((address) =>
+          isObject(allNftContracts[address]),
+        )
+      ) {
+        Object.keys(allNftContracts).forEach((address) => {
+          const nftContractsByChainId = allNftContracts[address];
+          if (isObject(nftContractsByChainId)) {
+            for (const chainId of Object.keys(nftContractsByChainId)) {
+              if (
+                !isStrictHexString(chainId) &&
+                nftContractsByChainId[toHex(chainId)]
+              ) {
+                delete nftContractsByChainId[chainId];
+              }
+            }
+          }
+        });
+      }
+    }
+
+    // Migrate NftController.allNfts
+    if (
+      hasProperty(nftControllerState, 'allNfts') &&
+      isObject(nftControllerState.allNfts)
+    ) {
+      const { allNfts } = nftControllerState;
+
+      if (Object.keys(allNfts).every((address) => isObject(allNfts[address]))) {
+        Object.keys(allNfts).forEach((address) => {
+          const nftsByChainId = allNfts[address];
+          if (isObject(nftsByChainId)) {
+            for (const chainId of Object.keys(nftsByChainId)) {
+              if (
+                !isStrictHexString(chainId) &&
+                nftsByChainId[toHex(chainId)]
+              ) {
+                delete nftsByChainId[chainId];
```

Comment:
[nit] Based on the condition just above it, do we want to delete `nftsByChainId[toHex(chainId)]` instead for this condition?

Not super familiar with the functionality here, we might need this: 
```js
if (
  !isStrictHexString(chainId) &&
  nftsByChainId[toHex(chainId)]
) {
  delete nftsByChainId[toHex(chainId)];
} else if (
  isStrictHexString(chainId) &&
  nftsByChainId[chainId]
){
  delete nftsByChainId[chainId];
}
```

in which case we can simplify

```js
const getHexChainId = (chainId) => !isStrictHexString(chainId) : toHex(chainId) : chainId;

const hexChainId = getHexChainId(chainId);
delete nftsByChainId[hexChainId]
```

Reply:
The purpose of this file is to delete duplicate entries in the ` allNfts[address]` objects. Because of the migration errors this PR addresses, a user could be in a situation where they have `allNfts[address].0x1` data and `allNfts[address].1` data. The latter would have been added after the failed migration, and the former would not have been visible in the UI since the failed migration. We want to delete the former, as they could overwrite the latter in migration 88, and  we assume that the tokens the user has seen in their wallet most recently are the tokens that they expect to see after this migration fix.